Remove deprecated methods from `ParserResult` and the `SingleSelectExecutor` class

mpdude opened the pull request:
This removes methods that have been deprecated in 2.20.x via #11188 and/or got deprecation notices added in #12196.

greg0ire reviewed (commented):
Please add something in UPGRADE.md

greg0ire commented:
The PHPStan baseline needs an update: https://github.com/doctrine/orm/blob/cdc52b2a9e07418538be6642c08232228ba47268/phpstan-baseline.neon#L2319-L2323

greg0ire commented on UPGRADE.md:
```diff
@@ -1,5 +1,9 @@
 # Upgrade to 4.0
 
+## BC BREAK: `\Doctrine\ORM\Query\ParserResult` no longer holds `AbstractSqlExecutor`, `Doctrine\ORM\Query\Exec\SingleSelectExecutor` is removed
```
You added a leading backslash on an occurrence, and not on another. Other entries don't include that leading backslash, so I'd remove it.

mpdude commented:
fixed

mpdude committed:
Remove deprecated methods from `ParserResult` and the `SingleSelectExecutor` class

This removes methods that have been deprecated in 2.20.x via #11188 and/or got deprecation notices added in #12196.

greg0ire commented:
Also, this needs a rebase.

derrabus commented on UPGRADE.md:
```diff
@@ -1,5 +1,16 @@
 
+## BC BREAK: Internal structure of `Doctrine\ORM\Query\ParserResult` changed
+
+The internal structure of `Doctrine\ORM\Query\ParserResult` changed. Thus, it is no longer possible to unserialize cached parser result instances that
+were created and serialized with the previous major version of the ORM.
```
Is it possible to create a compat layer?
